sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/PartitionBasedLoadBalancer.java

```diff
@@ -0,0 +1,331 @@
+        }
+
+        if (!isValid(partitionOwnershipMap)) {
+            // User data is corrupt.
+            return Mono.error(new IllegalStateException("Invalid partitionOwnership data from PartitionManager"));
+        }
+
+        /*
+         * Remove all partitions' ownership that have not be modified for a configuration period of time. This means
+         * that the previous EventProcessor that owned the partition is probably down and the partition is now eligible
+         * to be claimed by other EventProcessors.
+         */
+        Map<String, PartitionOwnership> activePartitionOwnershipMap = removeInactivePartitionOwnerships(
+            partitionOwnershipMap);
+
+        int numberOfPartitions = partitionIds.size();
+        if (ImplUtils.isNullOrEmpty(activePartitionOwnershipMap)) {
+            /*
+             * If the active partition ownership map is empty, this is the first time an event processor is
+             * running or all Event Processors are down for this Event Hub, consumer group combination. All
+             * partitions in this Event Hub are available to claim. Choose a random partition to claim ownership.
+             */
+            claimOwnership(partitionOwnershipMap, partitionIds.get(RANDOM.nextInt(numberOfPartitions)));
+            return Mono.empty();
+        }
+
+        /*
+         * Create a map of owner id and a list of partitions it owns
+         */
+        Map<String, List<PartitionOwnership>> ownerPartitionMap = activePartitionOwnershipMap.values()
+            .stream()
+            .collect(
+                Collectors.groupingBy(PartitionOwnership::ownerId, mapping(Function.identity(), toList())));
+
+        // add the current event processor to the map if it doesn't exist
+        ownerPartitionMap.putIfAbsent(this.ownerId, new ArrayList<>());
+
+        /*
+         * Find the minimum number of partitions every event processor should own when the load is
+         * evenly distributed.
+         */
+        int numberOfActiveEventProcessors = ownerPartitionMap.size();
+        int minPartitionsPerEventProcessor = numberOfPartitions / numberOfActiveEventProcessors;
+
+        /*
+         * If the number of partitions in Event Hub is not evenly divisible by number of active event processors,
+         * a few Event Processors may own 1 additional partition than the minimum when the load is balanced. Calculate
+         * the number of event processors that can own additional partition.
+         */
+        int numberOfEventProcessorsWithAdditionalPartition = numberOfPartitions % numberOfActiveEventProcessors;
+
+        if (isLoadBalanced(minPartitionsPerEventProcessor, numberOfEventProcessorsWithAdditionalPartition,
+            ownerPartitionMap)) {
+            // If the partitions are evenly distributed among all active event processors, no change required.
+            return Mono.empty();
```

Comment:
Lets say we have an Event Hubs with 3 partitions and we follows the below steps-
* Create Event Processor instance `processor1`
* call `processor1.start()`
* after some delay call `processor1.stop()`
* now  `processor1` owns all the 3 partitions and the load is balanced.
* if we call `processor1.start()` again, `isLoadBalanced()` will return true and here if we returns `Mono.empty()` it means we will stuck in the loop forever(if I understood correctly)?

 I think in this case we should create a new pump and start receiving events again from all the partitions?

Reply:
... or do we just reset the state to assume that things are not balanced each time we stop?   
... or do we assume things are balanced and then slowly realize they're not and re balance them?


I don't have a strong opinion, I think either case works - its just a matter of how aggressive we pursue work stealing.  However, I'd suggest that we avoid throwing away an instance and triggering an allocation for a new one if we just need a state reset.